reps/2023-01-09-aws-fleet-support.md

```diff
@@ -0,0 +1,147 @@
+
+### How EC2 fleet works?
+
+An EC2 Fleet contains configuration to launch a group of instances. There are [three types](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-fleet-request-type.html) of fleet requests, namely, `request`, `maintain` and `instant`. Because of the constraints mentioned in the **requirements** section (i.e., EC2 must not interfere with autoscaler activities), we would only be able to make use of `instant` type. Unlike, `maintain` and `request`, `instant` type `create_fleet` creates a fleet entity synchronously in an AWS account which is simply a configuration. Below are couple of observations related to `instant` type EC2 fleets:
+
+-   `instant` type fleets cannot be deleted without terminating instances it created.
+-   directly terminating instances created by `instant` fleet using `terminate_instances` API does not affect the fleet state.
+-   spot interruptions do not replace an instance created by `instant` fleet.
+-   there is no limit to number of EC2 `instant` type fleets one can have in an account.
+
+### How will EC2 fleets be created?
+
+An EC2 fleet configuration will be encapsulated within a single node type from the autoscaler perspective. We aim to create a new EC2 `instant` fleet as part of `create_node` method that is pretty much the entry point for any scale-up request from any other component (per diagram above). Ray's autoscaler will make sure `max_workers` config is honoured as well as perform any bin-packing of the resources across node types. EC2 automatically tags the instances created this way with a tag named `aws:ec2:fleet-id`.
+
+### How will the demand be converted into resources?
+
+Autoscaling monitor receives several demands from the metrics like vCPUs, memory, gpus etc. The [resource_demand_scheduler.py](https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/_private/resource_demand_scheduler.py) is responsible for converting the demand into number of workers per node type specified in the autoscaler config. However, it relies on the **node provider** to statically determine the CPU, memory, GPU and any custom resources. Hence, **node_provider** will determine CPU, memory and GPU from the `InstanceRequirements` and specified `InstanceType` parameters and aggregate them based on the low spec instance. Hence, autoscaler may end up spinning up more nodes than necessary which leads to overscaling. Autoscaler will also identify the idle nodes and remove them if required in the subsequent iterations. However, the [current behavior](https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/_private/aws/node_provider.py#L403) also does not guarantee target capacity will reached. Autoscaling monitor loop will figure out any underscaling issue
```

Comment:
The Ray scheduler will also use the resource capacities that are inferred here. I am not sure if that's desirable.
It may be possible to introduce some kludging to make sure the real resource capacities are perceived by the Ray internals.

Reply:
Are you suggesting resource demand scheduler to bin-pack based on real resource capacities instead of inferring it this way? If I am not wrong, the inherent assumption is that every node type will represent finite amount of resources (vCPU and memory) and hence it is being statically retrieved currently. We will only predict future capacity offered by fleets as there is no guarantee that EC2 will provision requested amount of units due to availability within instance pools.

Reply:
> The Ray scheduler will also use the resource capacities that are inferred here

I think that was only a legacy code path for handling the old `worker_nodes` field that was removed in 2.0. In any case, we probably shouldn't rely on that...

Reply:
here's the [code path which isn't active ](https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/_private/resource_demand_scheduler.py#L363)

Reply:
> I think that was only a legacy code path.

You might be thinking of something else. I'm referring to the `RAY_OVERRIDE_RESOURCES` env, which is used to advertise the (auto-filled) resources from the autoscaling config to the Ray scheduler.

If the proposal is implemented as is, then both the autoscaler and the Ray scheduler will see the reduced resource capacities. Come to think of it, I don't see a straightforward way of avoiding that here.

Side: There's a related contortion [here](https://github.com/ray-project/ray/blob/1a8f330ae9502b5911bc34cd34a42929a3d789d9/python/ray/autoscaler/_private/resource_demand_scheduler.py#L215) that will update _absent_ resources from runtime.

Reply:
Oh yeah I think you're right here. I think we should double down on only supporting node types with the same resource shape for now to avoid these problems for the most immediate proposal, and generally make sure the autoscaler is more well informed about node resources (rather than trying to guess what a crazy ray core heuristic will do) to avoid these issues if we need to broaden the class of node types supported by this.

Reply:
that sgtm

Reply:
Makes sense, I misunderstood the problem earlier. Resolving this thread.

Reply:
Probably the REP should updated to indicated that we will not initially recommend using instances with different resource capacities in a single fleet.

Reply:
Makes sense. Below is one of the item `Proposed code changes`:

> Update documentation to recommend developers to use instances with similar resource shape to avoid overscaling though EC2 fleet allows them to acquire instances at lower costs.

Do you want to callout explicitly in any other section?

Reply:
looks good